File: day06/aws-vpc-ts/privateSubnet.ts
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

interface CreatePrivateSubnetArgs {
  az: string;
  cidr: string;
  vpcId: pulumi.Input<string>;
  natGatewayId?: pulumi.Input<string>;
  defaultTags: Record<string, string>;
}

export class PrivateSubnet {
  public name: string;
  public subnet: aws.ec2.Subnet;

  private routeTable: aws.ec2.RouteTable;

  constructor(args: CreatePrivateSubnetArgs) {
    const subnet = new aws.ec2.Subnet(`my-private-subnet-${args.az}`, {
      vpcId: args.vpcId,
      cidrBlock: args.cidr,
      availabilityZone: args.az,
      tags: {
        "Name": `my-private-subnet-${args.az}`,
        ...args.defaultTags
      }
    });

    const rt = new aws.ec2.RouteTable(`my-private-subnet-${args.az}-rt`, {
      vpcId: args.vpcId,
      tags: {
        'Name': `my-private-subnet-${args.az}-rt`,
        ...args.defaultTags
      }
    });

    new aws.ec2.RouteTableAssociation(`my-private-subnet-${args.az}-rt-association`, {
      routeTableId: rt.id,
      subnetId: subnet.id,
    });

    this.name = `my-private-subnet-${args.az}`;
    this.subnet = subnet;
    this.routeTable = rt;
  }

  addNatGateway(natGatewayId: pulumi.Input<string>) {
    new aws.ec2.Route(`${this.name}-nat-gateway-route`, {
      routeTableId: this.routeTable.id,
      destinationCidrBlock: '0.0.0.0/0',
      natGatewayId: natGatewayId,
    });
  }
}

File: day06/day06-aws-vpc-cs/day06-aws-vpc-cs/PrivateSubnet.cs
using Pulumi;
using Pulumi.Aws.Ec2;
using Aws = Pulumi.Aws;

namespace day06_aws_vpc_cs
{
    public class PrivateSubnetArgs
    {
        public Input<string> VpcId { get; set; }
        public string Cidr { get; set; }
        public Input<string> Az { get; set; }
        public InputMap<string> Tags { get; set; }
    }

    public class PrivateSubnet
    {
        private string Name { get; set; }
        public Subnet Subnet { get; set; }

        public RouteTable RouteTable { get; set; }

        public PrivateSubnet(string name, PrivateSubnetArgs args)
        {
            Name = name;

            var subnetName = args.Az.Apply(az => "my-private-subnet-" + az);

            Subnet = new Subnet(Name, new SubnetArgs()
            {
                VpcId = args.VpcId,
                CidrBlock = args.Cidr,
                AvailabilityZone = args.Az,
                Tags = InputMap<string>.Merge(args.Tags, new InputMap<string>
                {
                    { "Name", subnetName }
                })
            });

            RouteTable = new RouteTable(this.Name, new RouteTableArgs
            {
                VpcId = args.VpcId,
                Tags = InputMap<string>.Merge(args.Tags, new InputMap<string>
                {
                    { "Name", subnetName.Apply(n => n + "-rt") }
                })
            });

            new Aws.Ec2.RouteTableAssociation(this.Name, new RouteTableAssociationArgs
            {
                RouteTableId = RouteTable.Id,
                SubnetId = Subnet.Id
            });
        }

        public void AddNatGateway(Input<string> natGatewayId)
        {
            new Aws.Ec2.Route(Name + "-nat-gateway-route", new RouteArgs
            {
                DestinationCidrBlock = "0.0.0.0/0",
                RouteTableId = RouteTable.Id,
                NatGatewayId = natGatewayId
            });
        }
    }
}